### ISSUE TITLE
Distinctness proof

### ISSUE THREAD

**Issue by freebroccolo** [freebroccolo/agda-cubical-sets#10]
Here's a complete version that defines the relation directly on the derivations:

```agda
module Wellfounded where

data _≡_ {A : Set} (a : A) : A → Set where
  refl : a ≡ a

data Bool : Set where
  #0 : Bool
  #1 : Bool

data _≅_ : Bool → Bool → Set where
  idn : ∀ {a b}   (p : a ≡ b) → a ≅ b
  seq : ∀ {a b} x (p : a ≅ x) (q : x ≅ b) → a ≅ b
  inv : ∀ {a b}   (p : b ≅ a) → a ≅ b

data _<_ : ∀ {a b c d} → a ≅ b → c ≅ d → Set where
  stop-seq₀ : ∀ {a b c} (p : a ≅ b) (q : b ≅ c) → p < seq b p q
  stop-seq₁ : ∀ {a b c} (p : a ≅ b) (q : b ≅ c) → q < seq b p q
  stop-inv  : ∀ {a b} (p : a ≅ b) → p < inv p
  next-seq₀ : ∀ {a b c d e} (p : a ≅ b) (q : c ≅ d) (r : d ≅ e) → p < q → p < seq d q r
  next-seq₁ : ∀ {a b c d e} (p : a ≅ b) (q : c ≅ d) (r : d ≅ e) → p < r → p < seq d q r
  next-inv  : ∀ {a b c d} (p : a ≅ b) (q : c ≅ d) → p < q → p < inv q

data Acc {a b} (p : a ≅ b) : Set where
  acc : (∀ {c}{d} (q : c ≅ d) → q < p → Acc q) → Acc p

mutual
  wf : ∀ {a b} (p : a ≅ b) → Acc p
  wf (idn refl) = acc λ {q ()}
  wf (seq x p q) = acc (wf-seq p q)
  wf (inv p) = acc (wf-inv p)

  wf-inv : ∀ {a b c d} (p : a ≅ b) (q : c ≅ d) → q < inv p → Acc q
  wf-inv p .p (stop-inv .p) = wf p
  wf-inv (idn refl) q (next-inv .q .(idn refl) ())
  wf-inv (seq x p₀ p₁) q (next-inv .q .(seq x p₀ p₁) φ) = wf-seq p₀ p₁ q φ
  wf-inv (inv p) q (next-inv .q .(inv p) φ) = wf-inv p q φ

  wf-seq : ∀ {a b c d e} (p : a ≅ b) (q : b ≅ c) (r : d ≅ e) → r < seq b p q → Acc r
  wf-seq p q .p (stop-seq₀ .p .q) = wf p
  wf-seq p q .q (stop-seq₁ .p .q) = wf q
  wf-seq (idn refl) q r (next-seq₀ .r .(idn refl) .q ())
  wf-seq (seq x p₀ p₁) q r (next-seq₀ .r .(seq x p₀ p₁) .q φ) = wf-seq p₀ p₁ r φ
  wf-seq (inv p) q r (next-seq₀ .r .(inv p) .q φ) = wf-inv p r φ
  wf-seq p (idn refl) r (next-seq₁ .r .p .(idn refl) ())
  wf-seq p (seq x q₀ q₁) r (next-seq₁ .r .p .(seq x q₀ q₁) φ) = wf-seq q₀ q₁ r φ
  wf-seq p (inv q) r (next-seq₁ .r .p .(inv q) φ) = wf-inv q r φ

data ⊥ : Set where

mutual
  distinct-acc-fwd : (p : #0 ≅ #1) → Acc p → ⊥
  distinct-acc-fwd (idn ()) φ
  distinct-acc-fwd (seq #0 p₀ p₁) (acc φ) = distinct-acc-fwd p₁ (φ p₁ (stop-seq₁ p₀ p₁))
  distinct-acc-fwd (seq #1 p₀ p₁) (acc φ) = distinct-acc-fwd p₀ (φ p₀ (stop-seq₀ p₀ p₁))
  distinct-acc-fwd (inv p) (acc φ) = distinct-acc-bwd p (φ p (stop-inv p))

  distinct-acc-bwd : (p : #1 ≅ #0) → Acc p → ⊥
  distinct-acc-bwd (idn ()) φ
  distinct-acc-bwd (seq #0 p₀ p₁) (acc φ) = distinct-acc-bwd p₀ (φ p₀ (stop-seq₀ p₀ p₁))
  distinct-acc-bwd (seq #1 p₀ p₁) (acc φ) = distinct-acc-bwd p₁ (φ p₁ (stop-seq₁ p₀ p₁))
  distinct-acc-bwd (inv p) (acc φ) = distinct-acc-fwd p (φ p (stop-inv p))

distinct : #0 ≅ #1 → ⊥
distinct p = distinct-acc-fwd p (wf p)
```

**Comment by freebroccolo** [comment]
It might also be cleaner to just skip the stuff with `depth` and `max` and to define `_<_` directly on `_≅_`.

**Comment by freebroccolo** [comment]
Here's a complete version that defines the relation directly on the derivations:

```
module Wellfounded where

data _≡_ {A : Set} (a : A) : A → Set where
  refl : a ≡ a

data Bool : Set where
  #0 : Bool
  #1 : Bool

data _≅_ : Bool → Bool → Set where
  idn : ∀ {a b}   (p : a ≡ b) → a ≅ b
  seq : ∀ {a b} x (p : a ≅ x) (q : x ≅ b) → a ≅ b
  inv : ∀ {a b}   (p : b ≅ a) → a ≅ b

data _<_ : ∀ {a b c d} → a ≅ b → c ≅ d → Set where
  stop-seq₀ : ∀ {a b c} (p : a ≅ b) (q : b ≅ c) → p < seq b p q
  stop-seq₁ : ∀ {a b c} (p : a ≅ b) (q : b ≅ c) → q < seq b p q
  stop-inv  : ∀ {a b} (p : a ≅ b) → p < inv p
  next-seq₀ : ∀ {a b c d e} (p : a ≅ b) (q : c ≅ d) (r : d ≅ e) → p < q → p < seq d q r
  next-seq₁ : ∀ {a b c d e} (p : a ≅ b) (q : c ≅ d) (r : d ≅ e) → p < r → p < seq d q r
  next-inv  : ∀ {a b c d} (p : a ≅ b) (q : c ≅ d) → p < q → p < inv q

data Acc {a b} (p : a ≅ b) : Set where
  acc : (∀ {c}{d} (q : c ≅ d) → q < p → Acc q) → Acc p

mutual
  wf : ∀ {a b} (p : a ≅ b) → Acc p
  wf (idn refl) = acc λ { q ()}
  wf (seq x p q) = acc (wf-seq p q)
  wf (inv p) = acc (wf-inv p)

  wf-inv : ∀ {a b c d} (p : a ≅ b) (q : c ≅ d) → q < inv p → Acc q
  wf-inv p .p (stop-inv .p) = wf p
  wf-inv (idn refl) q (next-inv .q .(idn refl) ())
  wf-inv (seq x p₀ p₁) q (next-inv .q .(seq x p₀ p₁) φ) = wf-seq p₀ p₁ q φ
  wf-inv (inv p) q (next-inv .q .(inv p) φ) = wf-inv p q φ

  wf-seq : ∀ {a b c d e} (p : a ≅ b) (q : b ≅ c) (r : d ≅ e) → r < seq b p q → Acc r
  wf-seq p q .p (stop-seq₀ .p .q) = wf p
  wf-seq p q .q (stop-seq₁ .p .q) = wf q
  wf-seq (idn refl) q r (next-seq₀ .r .(idn refl) .q ())
  wf-seq (seq x p₀ p₁) q r (next-seq₀ .r .(seq x p₀ p₁) .q φ) = wf-seq p₀ p₁ r φ
  wf-seq (inv p) q r (next-seq₀ .r .(inv p) .q φ) = wf-inv p r φ
  wf-seq p (idn refl) r (next-seq₁ .r .p .(idn refl) ())
  wf-seq p (seq x q₀ q₁) r (next-seq₁ .r .p .(seq x q₀ q₁) φ) = wf-seq q₀ q₁ r φ
  wf-seq p (inv q) r (next-seq₁ .r .p .(inv q) φ) = wf-inv q r φ

data ⊥ : Set where

mutual
  distinct-acc-fwd : (p : #0 ≅ #1) → Acc p → ⊥
  distinct-acc-fwd (idn ()) φ
  distinct-acc-fwd (seq #0 p₀ p₁) (acc φ) = distinct-acc-fwd p₁ (φ p₁ (stop-seq₁ p₀ p₁))
  distinct-acc-fwd (seq #1 p₀ p₁) (acc φ) = distinct-acc-fwd p₀ (φ p₀ (stop-seq₀ p₀ p₁))
  distinct-acc-fwd (inv p) (acc φ) = distinct-acc-bwd p (φ p (stop-inv p))

  distinct-acc-bwd : (p : #1 ≅ #0) → Acc p → ⊥
  distinct-acc-bwd (idn ()) φ
  distinct-acc-bwd (seq #0 p₀ p₁) (acc φ) = distinct-acc-bwd p₀ (φ p₀ (stop-seq₀ p₀ p₁))
  distinct-acc-bwd (seq #1 p₀ p₁) (acc φ) = distinct-acc-bwd p₁ (φ p₁ (stop-seq₁ p₀ p₁))
  distinct-acc-bwd (inv p) (acc φ) = distinct-acc-fwd p (φ p (stop-inv p))

distinct : #0 ≅ #1 → ⊥
distinct p = distinct-acc-fwd p (wf p)
```

**Comment by jonsterling** [comment]
wow!

**Comment by freebroccolo** [comment]
After I implemented this I felt kind of dumb because I realized that you don't actually need the accessibility proofs anymore since the equational language has been simplified too much. I'm pretty certain it will still work fine once all the other cases are added back for the full De Morgan algebras though (and then some of the recursive calls will definitely need this for termination).

But I also had another idea that might be even cleaner than this. Instead of defining `_<_` and proving well-foundedness, we could probably just define a separate set of derivations, i.e.,

```agda
data _≇_ : Bool → Bool → Set where
```

and an interpretation function:

```agda
⟦_⟧ : ∀ {a b} → a ≇ b → a ≅ b → ⊥
```

I've done stuff like this before when I wanted proofs I could easily inspect and manipulate (instead of functions) but haven't tried it for termination. It looks like it should work here since we already have this kind of building up of a structure (the "fuel" for the accessibility proofs I guess) and subsequent interpretation (consuming the "fuel" to make another step in the call graph).

This would change the types for `distinct-acc-fwd` and `distinct-acc-bwd`:

```agda
distinct-acc-fwd : #0 ≇ #1
distinct-acc-bwd : #1 ≇ #0
```

The actual definition of `_≇_` would look a lot like `_<_` does now.
